core/src/main/java/org/elasticsearch/index/fieldvisitor/CustomFieldsVisitor.java

```diff
@@ -43,6 +43,6 @@ public Status needsField(FieldInfo fieldInfo) throws IOException {
             return Status.YES;
         }
 
-        return fields.contains(fieldInfo.name) ? Status.YES : Status.NO;
+        return fields == null ? Status.NO : fields.contains(fieldInfo.name) ? Status.YES : Status.NO;
```

Comment:
Why do we need the null check? Should we requires that `fields` is not null instead and fix callers to pass `Collections.emptySet()` when they don't want to load any field?